design/UsageValidationRules.md

```diff
@@ -0,0 +1,82 @@
+
+## Render passes
+
+In render passes the only writable resources are textures used as `OUTPUT_ATTACHMENT` and resources used as `STORAGE`.
+
+To avoid data hazards the simplest validation rules would be to check for every subresource that it is used as either:
+ - `OUTPUT_ATTACHMENT`
+ - A combination of read-only usages
+ - `STORAGE`
+
+If there is no usage of `STORAGE` then there are no data-races as everything is read-only, except `OUTPUT_ATTACHMENT` which is well-ordered.
+For `STORAGE` resources, the order of reads and writes to the same memory location is undefined and it is up to the application to ensure data-race freeness.
+
+## Compute passes
+
+They are similar to render passes, every subresource (alternatively every resource) must be used as either:
+ - A combination of read-only usages
+ - `STORAGE`
+
+In a single `dispatch` command, the order of reads and writes to the same memory location of a `STORAGE` resource is undefined.
+It is up to the application to ensure data-race freeness.
+
+## Other operations
+
+Assuming we don't have "copy passes" and that other operations are top-level, then a command buffers looks like a sequence of:
+ - Compute passes
+ - Render passes
+ - Copies and stuff
+
+Implementations ensure that there are no-data races between any of these operations.
+There are validation rules to check for example that the source and destinations regions of buffer copies don't have overlap.
+
+## Discussion
+
+In this proposal the only sources of data races come from the `STORAGE` usage:
+ - In render passes inside a single drawcall and between drawcalls for accesses to memory written to as `STORAGE`.
+ - In any compute passes inside a single dispatch for accesses to memory written to as `STORAGE`
+ - In `manualStorageBarrier` compute passes (see below) between dispatches for accesses to memory written to as `STORAGE`
+
+There are cases where the resource usage tracking could get expensive for example if subresources of a texture with an odd index use `STORAGE` while even ones use `SAMPLED`.
+That said such a case is unlikely to happen in real-world applications.
+
+The cost of the state tracking necessary to add memory barriers inside compute passes and at command buffer top-level seems manageable (though difficult to describe in the 30 minutes I have before the meeting).
+
+There is no read-only storage that could be used in combination with other read-only usages because some APIs don't support it.
+For example D3D12 assumes UAV is always writeable and disallows transitioning to a combination of UAV and a read-only usage.
+
+## Open questions
+
+### More constrained texture usage validation
+
+Each layer and mip-level of textures can have an independent usage which means implementations might need to track usage per mip-level per layer of a resource.
+If this is deemed too costly, we could only have two sub-resources tracked in textures: the part as `OUTPUT_ATTACHMENT` and the rest.
```

Comment:
There are uses cases (VR/AR being one of them) where developers will want to two subresources to be `OUTPUT_ATTACHMENT` at the same time.  

I'm not overly worried about the case you described above where developers have one resource in the `STORAGE` and the other one in `SAMPLED`.  I think what will be more common is for some subresources to be in `TRANSFER_DST` while others are in `SAMPLED` state.  Will this latter case require "expensive" resource tracking like the former one will?  

Forcing developers to have all of their mips accounted for before being able to sample may be fine for an MVP.  But, we should consider opening up this specific use case (`TRANSFER_DST`+`SAMPLED`) at a later point.

Reply:
`TRANSFER_DST` and `SAMPLED` happen outside of passes and inside of passes respectively so they are easier to track. There is no need to account for all mip since there is no concept of memory barrier / usage transition. However implementations will have to track inside a pass which mips are used as `STORAGE` and which are used as something else.